### DISCUSSION TITLE
Trying to understand style tables and responding to key events

### DISCUSSION THREAD

**Discussion by diego-crespo** [fltk-rs/fltk-rs#647] [General]
First I would like to say I appreciate the work that has gone into the fltk-rs library. I am still very new to Rust, and I've found it the easiest to understand of all the current Rust GUI libraries. It also has so far worked perfectly on my Raspberry Pi 4 where I'm doing my testing, while I have struggled to get rust guis working on it. So thank you Mohammed Alyousef! 

Now for my question. I don't understand how to style the text in the TextBuffer of the TextDisplay. I've seen the [Erco's Cheatsheet example ](http://seriss.com/people/erco/fltk/#Fl_Text_Display) but I don't understand C++ so I don't get how 

`     sbuff->text("AAAAAAAAAA\nBBBBBBBBBB\nCCCCCCCCCC\nDDDDDDDDDD\n"
                 "AAAAAAAAAA\nBBBBBBBBBB\nCCCCCCCCCC\nDDDDDDDDDD\n");`

Makes each line a different color. In my example below I'd like to change the "Hello" in Hello World to green, and the "World" to Red. 

```
use fltk::{app::*, window::*, prelude::*, frame::*, enums::*, text::*, input::*, text::*};

fn main() {
    let app = App::default().with_scheme(Scheme::Gleam);
    let mut win = Window::default()
	.with_label("A regular day")
	.with_size(500,500)
	.center_screen();
    let mut buf = TextBuffer::default();
    let mut sbuf = TextBuffer::default();
    let styles: Vec<StyleTableEntry> = vec![
	StyleTableEntry {
	    color: Color::Green,
            font: Font::Courier,
	    size: 16,
	},
       StyleTableEntry {
           color: Color::Red,
           font: Font::Courier,
	   size: 16
       }
    ];   
    let mut text_dis = TextDisplay::default()
	.with_size(100,100)
	.center_of(&win);

    text_dis.set_buffer(Some(buf));
    text_dis.set_highlight_data(sbuf.clone(), styles);
    text_dis.buffer().unwrap().append("Hello World");
    
    win.end();
    win.show();

    win.handle(move |_, event| {
	match event {
	    Event::KeyUp => {
		println!("A Key Up");
		text_dis.set_pos(text_dis.x(), text_dis.y() + 5);
		    true},
	    _ => false
	}
    });

    app.run().unwrap();
}
```
Lastly I don't think I totally understand how matching key events works. I have logic in the match to handle a `Event::KeyUp` but every button I press prints "A Key Up". I would expect this to only match the KeyUp key on my keyboard. Also setting the TextDisplay to a different position based on the key press doesn't seem to work  unless I add a` text_dis.redraw() `after the `set_pos` line. Unfortunately this causes multiple windows to be drawn on top of each other offset by 5px. How do I redraw a TextDisplay after I've changed the positions while clearing data related to the old position. 

In summary
1. How to I use style buffers to individually highlight words in a TextDisplay
2. How do I get my window handle function to print KeyUp on a KeyUp event
3. How do move a TextDisplay while clearing the data for the old one, so it doesn't look like it's just being smeared across the screen everytime the Event::KeyUp key is pressed?

**Comment by MoAlyousef** [comment]
Thank you!

1 - You append to the style buffer the length of the string you want to style. The alphabet characters represent the index in the style table, and repetition is the number of characters to apply that style to. So for example coloring hello world:
```rust
    text_dis.buffer().unwrap().append("Hello World");
    sbuf.append(&"A".repeat("Hello".len()));
    sbuf.append(" ");
    sbuf.append(&"B".repeat("World".len()));
```
It's not the most ergonomic but it's FLTK's way of doing things!

2/3 - The KeyUp Event is when a key goes up, you want the Key::Up key :)
Also you'll need to redraw the parent itself to avoid the smearing effect of movement.
```rust
    set_focus(&win);

    win.handle(move |w, event| match event {
        Event::KeyDown => {
            if event_key() == Key::Up {
                println!("A Key Up");
                text_dis.set_pos(text_dis.x(), text_dis.y() + 5);
                w.redraw();
                true
            } else {
                false
            }
        }
        _ => false,
    });
```
We use set focus because by text_default the display will take the focus in this example, otherwise, you can handle the Key event for the text_display instead:
```rust
    text_dis.handle(move |dis, event| match event {
        Event::KeyDown => {
            if event_key() == Key::Up {
                println!("A Key Up");
                dis.set_pos(dis.x(), dis.y() + 5);
                dis.parent().unwrap().redraw();
                true
            } else {
                false
            }
        }
        _ => false,
    });
```

**Reply by diego-crespo** [reply]
Oh that is embarrassing I should have realized I was using the wrong event! Thank you everyone works perfectly now
